lib/Command/UpsertProvider.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * @copyright Copyright (c) 2021 Bernd Rederlechner tsdicloud@github.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\UserOIDC\Command;
+
+use OC\Core\Command\Base;
+
+use OCA\UserOIDC\Db\ProviderMapper;
+
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class UpsertProvider extends Base {
+	private $providerMapper;
+
+	public function __construct(ProviderMapper $providerMapper) {
+		parent::__construct();
+		$this->providerMapper = $providerMapper;
+	}
+
+	protected function configure() {
+		$this
+			->setName('user_oidc:provider')
+			->setDescription('Create, show or update a OpenId connect provider connfig given the identifier of a provider')
+			->addArgument('providerid', InputOption::VALUE_REQUIRED, 'Administrative identifier name of the provider in the setup')
+			->addOption('clientid', 'c', InputOption::VALUE_REQUIRED, 'OpenID client identifier')
+			->addOption('clientsecret', 's', InputOption::VALUE_REQUIRED, 'OpenID client secret')
+			->addOption('discoveryuri', 'd', InputOption::VALUE_REQUIRED, 'OpenID discovery endpoint uri');
+		parent::configure();
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$providerid = $input->getArgument('providerid');
+		$clientid = $input->getOption('clientid');
+		$clientsecret = $input->getOption('clientsecret');
+		$discoveryuri = $input->getOption('discoveryuri');
```

Comment:
for type safety you might want to cast the four input arguments/options here to their respective types

Reply:
This is not so easy because of the optionl handling of the fields on updates. Tried it out and it worked not so well. All types are strings and the createOrupdate signature is typed, so I would assume implicit casting is enough.